health/health.go

```diff
@@ -0,0 +1,82 @@
+package health
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/featureform/metadata"
+	"github.com/featureform/provider"
+	pc "github.com/featureform/provider/provider_config"
+	pt "github.com/featureform/provider/provider_type"
+)
+
+const baseInitializationErrorMessage = "Featureform failed to connect to provider with the supplied credentials due to the following error:"
+
+type Health struct {
+	metadata *metadata.Client
+}
+
+func NewHealth(client *metadata.Client) *Health {
+	return &Health{
+		metadata: client,
+	}
+}
+
+func (h *Health) Check(providerName string) (bool, error) {
+	rec, err := h.metadata.GetProvider(context.Background(), providerName)
+	if err != nil {
+		return false, err
+	}
+	if !h.isSupportedProvider(pt.Type(rec.Type())) {
+		return false, fmt.Errorf("unsupported provider type: %s", rec.Type())
+	}
+	p, err := provider.Get(pt.Type(rec.Type()), pc.SerializedConfig(rec.SerializedConfig()))
+	if err != nil {
+		return false, errors.New(h.handleError(err))
+	}
+	isHealthy, err := p.Check()
+	if err != nil {
+		return false, errors.New(h.handleError(err))
+	}
+	return isHealthy, nil
+}
+
+func (h *Health) isSupportedProvider(t pt.Type) bool {
+	switch t {
+	case pt.RedisOnline, pt.DynamoDBOnline, pt.PostgresOffline, pt.SnowflakeOffline, pt.SparkOffline:
+		return true
+	default:
+		return false
+	}
+}
+
+func (h *Health) handleError(err error) string {
```

Comment:
I am wondering if it makes sense to pass in the provider type and run the general checks and only checks for that provider? Then we can move the individual provider error checks into separate functions. If you agree, we can add it as a TODO, doesn't have to be done in this PR.

Reply:
I mostly agree with the above. We can have a general error handler but I think its cleaner if each provider is responsible for handling its own errors. The below can get really messy

Reply:
I originally wrote this pattern until I started to see enough overlap to second guess myself. I've now added @sdreyer's recommendation for typed errors and am switching on these for Spark. We can add others if necessary.